### PR TITLE
Improve workload execution states

### PR DESCRIPTION
**Pull request by krucod3**
Issues: #165

# Definition of Done

The PR shall be merged only if all items mentioned in [CONTRIBUTING.md](https://github.com/eclipse-ankaios/ankaios/blob/main/CONTRIBUTING.md#how-to-contribute) have been followed. In case an item is not applicable as described, please provide a short explanation in the description.

- [ ] All steps in [CONTRIBUTING.md](https://github.com/eclipse-ankaios/ankaios/blob/main/CONTRIBUTING.md#how-to-contribute) have been handled

### CODE REVIEW

**Comment by krucod3** [comment]
This is the initial draft with changes in proto which are still not implemented in the code (everything is broken).

File: api/proto/ankaios.proto [hunk 1]
```diff
@@ -130,29 +130,81 @@ enum AddCondition {
   ADD_COND_FAILED = 2; /// The workload has exited with an error or could not be started.
 }
 
+
+message ExecutionState {
+    string additionalInfo = 1;
```
Comment (maturar) [inline api/proto/ankaios.proto:135]: Is there any description for the `additionalInfo`?

File: api/proto/ankaios.proto [hunk 2]
```diff
@@ -130,29 +130,81 @@ enum AddCondition {
...
+enum Succeeded {
+    SUCCEEDED_OK = 0; /// The workload has successfully finished operation.
+    SUCCEEDED_DELETE_FAILED = 8; /// The deletion of the workload by the runtime failed.
+}
+
+/**
+* The workload has failed or is in a degraded state.
+*/
+enum Failed {
+    FAILED_NOK = 0; /// The workload has failed during operation
```
Comment (maturar) [inline api/proto/ankaios.proto:189]: What about calling this `FAILED_EXEC_FAILED`? The "NOK" seems to me too general.

File: common/doc/swdesign/drawio/state_workload_execution_all_states.drawio.svg
Comment (maturar) [inline common/doc/swdesign/drawio/state_workload_execution_all_states.drawio.svg:1]: Is there a description of the colors used in this diagram (a legend) somewhere?

File: api/proto/ankaios.proto (hunk 1 above)
Comment (krucod3) [reply api/proto/ankaios.proto:135]:
```suggestion
    string additionalInfo = 1; /// The additional info contains more detailed information from the runtime regarding the execution state.
```

File: common/doc/swdesign/drawio/state_workload_execution_all_states.drawio.svg
Comment (krucod3) [reply common/doc/swdesign/drawio/state_workload_execution_all_states.drawio.svg:1]: I added one already.

File: api/proto/ankaios.proto (hunk 2 above)
Comment (krucod3) [reply api/proto/ankaios.proto]: Done.

File: ank/src/cli_commands.rs [hunk 3]
```diff
@@ -650,19 +653,22 @@ mod tests {
                 name: String::from("name1"),
                 agent: String::from("agent_A"),
                 runtime: String::from("runtime"),
-                execution_state: String::from("Running"),
+                execution_state: ExecutionState::running().state.to_string(),
+                additional_info: String::from(""),
```
Comment (inf17101) [inline ank/src/cli_commands.rs:657]:
```suggestion
                additional_info: Default:default(),
```

oder 

```suggestion
                ..Default::default(),
```

File: common/src/objects/workload_state.rs [hunk 4]
```diff
@@ -18,82 +18,360 @@ use serde::{Deserialize, Serialize};
...
-}
-
-impl From<i32> for ExecutionState {
+use crate::std_extensions::UnreachableOption;
+
+use super::WorkloadExecutionInstanceName;
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum PendingSubstate {
+    Initial = 0,
```
Comment (inf17101) [inline common/src/objects/workload_state.rs:27]: I cannot see that conversion to `Initial` is unit tested and that the state is set to initial in production code at the beginning.

File: common/doc/swdesign/README.md [hunk 5]
```diff
@@ -121,6 +132,45 @@ The Following diagram shows all Ankaios workload states and the possible transit
...
 
+#### Ankaios workload execution state additional information
+`swdd~common-workload-state-additional-information~1`
+
+Status: approved
+
+Ankaios shall support additional information for the workload execution state.
+
+Rationale:
+The additional information is could be provided by the runtime and is helpful for debugging purposes.
```
Comment (lingnoi) [inline common/doc/swdesign/README.md:143]:
```suggestion
The additional information could be provided by the runtime and is helpful for debugging purposes.
```

File: server/src/ankaios_server.rs [hunk 6]
```diff
@@ -987,13 +968,17 @@ mod tests {
         assert!(comm_middle_ware_receiver.try_recv().is_err());
     }
 
+
+    // TODO the following test tests also the workload state db
+    // This should be avoided, the tests should be 
+
```
Comment (inf17101) [inline server/src/ankaios_server.rs:974]: Shall this be done in separate PR or is it missing inside this PR?

File: common/src/objects/workload_state.rs [hunk 7]
```diff
@@ -18,82 +18,360 @@ use serde::{Deserialize, Serialize};
...
+            }
+            proto::execution_state::ExecutionStateEnum::Removed(_) => ExecutionStateEnum::Removed,
+        }
+    }
+}
+
+// [impl->swdd~common-workload-states-supported-states~1]
+#[derive(Default, Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(default, rename_all = "camelCase")]
+
```
Comment (lingnoi) [inline common/src/objects/workload_state.rs:213]:
```suggestion
```

File: server/src/workload_state_db.rs [hunk 8]
```diff
@@ -30,85 +33,49 @@ impl WorkloadStateDB {
...
-                    execution_state: state.clone(),
-                })
-                .collect();
-            all_workload_states.append(&mut x);
-        }
-
-        all_workload_states
+        self.stored_states
+            .iter()
+            .flat_map(|(_, v)| v.iter().flat_map(|(_, v)| v.to_owned()))
```
Comment (lingnoi) [inline server/src/workload_state_db.rs:38]:
```suggestion
            .flat_map(|(_, workload_states)| workload_states.iter().flat_map(|(_, workload_state)| workload_state.to_owned()))
```

File: server/src/workload_state_db.rs [hunk 9]
```diff
@@ -30,85 +33,49 @@ impl WorkloadStateDB {
...
-                }
-            }
-        }
-
-        output_workload_states
+        self.stored_states
+            .iter()
+            .filter(|(k, _)| *k != excluding_agent_name)
+            .flat_map(|(_, v)| v.iter().flat_map(|(_, v)| v.to_owned()))
+            .collect()
```
Comment (lingnoi) [inline server/src/workload_state_db.rs:57]:
```suggestion
            .filter(|(agent_name, _)| *agent_name != excluding_agent_name)
            .flat_map(|(_, workload_states)| workload_states.iter().flat_map(|(_, workload_state)| workload_state.to_owned()))
            .collect()
```

File: server/src/workload_state_db.rs [hunk 10]
```diff
@@ -15,8 +15,11 @@
...
 
-type WorkloadStatesMap = HashMap<String, common::objects::ExecutionState>;
-type AgentWorkloadStates = HashMap<String, WorkloadStatesMap>;
+type WorkloadName = String;
+type AgentName = String;
+
+type WorkloadStatesMap = HashMap<WorkloadName, Vec<WorkloadState>>;
+type AgentWorkloadStates = HashMap<AgentName, WorkloadStatesMap>;
 
 pub struct WorkloadStateDB {
```
Comment (inf17101) [inline server/src/workload_state_db.rs:24]: For this module only the insert method is linked. No utest or other public method is linked at all.

File: common/doc/swdesign/README.md [hunk 11]
```diff
@@ -121,6 +132,45 @@ The Following diagram shows all Ankaios workload states and the possible transit
 
 ![Workload states](plantuml/state_workload_execution_states.svg)
 
+#### Ankaios workload execution state additional information
+`swdd~common-workload-state-additional-information~1`
+
+Status: approved
+
+Ankaios shall support additional information for the workload execution state.
```
[the remark this answers is not in the document]
Comment (krucod3) [reply common/doc/swdesign/README.md:140]:
```suggestion
Ankaios shall support a string with additional information for the workload execution state.
```

File: server/src/ankaios_server.rs (hunk 6 above)
Comment (krucod3) [reply server/src/ankaios_server.rs:974]: A separate one
```suggestion
```

File: agent/src/runtime_connectors/podman/podman_runtime.rs [hunk 12]
```diff
@@ -498,14 +504,14 @@ mod tests {
...
         let workload_id = PodmanWorkloadId {
             id: "test_id".into(),
         };
         let checker = PodmanStateGetter {};
         let res = checker.get_state(&workload_id).await;
-        assert_eq!(res, ExecutionState::ExecRunning);
+        assert_eq!(res, ExecutionState::running());
     }
 
     // [utest->swdd~podman-state-getter-returns-removed-state~1]
```
[the remark this answers is not in the document]
Comment (inf17101) [reply agent/src/runtime_connectors/podman/podman_runtime.rs:517]: I have changed the requirement content and the sequence diagram. Can be reviewed by another one.

File: server/src/ankaios_server.rs [hunk 13]
```diff
@@ -987,13 +968,17 @@ mod tests {
...
     // [utest->swdd~server-uses-async-channels~1]
     // [utest->swdd~server-stores-workload-state~1]
     // [utest->swdd~server-set-workload-state-unknown-on-disconnect~1]
```
[the remark this answers is not in the document]
Comment (inf17101) [reply server/src/ankaios_server.rs]: I have committed the adapted requirement and linked again.

File: tests/resources/ankaios.resource [hunk 14]
```diff
@@ -173,6 +173,19 @@ the workload "${workload_name}" shall not exist
...
 
+the workload "${workload_name}" shall not exist on agent "${expected_agent_name}" within "${timeout}" seconds
+    @{list_result}=    Create List
+    IF    ${timeout} > 0
+        @{list_result}=    Wait For Workload Removal
+        ...    %{ANK_BIN_DIR}${/}ank get workloads
+        ...    ${workload_name}
+        ...    ${expected_agent_name}
+        ...    ${timeout}
+    END
```
Comment (inf17101) [inline tests/resources/ankaios.resource:184]: I see that we used that pattern more often in this file, but in general if someone uses that keyword and passes timeout 0 then the underlying `ank get workloads` is never executed, but the assertion is green (Should be empty list). Then we have a false-positive stest. I think the other occurrences do not have this problem because they do the reverse check on the list or use the timeout different.

File: agent/src/workload/workload_control_loop.rs [hunk 15]
```diff
@@ -148,9 +148,9 @@ impl WorkloadControlLoop {
             control_loop_state
                 .update_state_tx
                 .update_workload_state(vec![common::objects::WorkloadState {
-                    agent_name: control_loop_state.instance_name.agent_name().into(),
-                    workload_name: control_loop_state.instance_name.workload_name().into(),
-                    execution_state: ExecutionState::ExecFailed,
+                    instance_name: control_loop_state.instance_name.to_owned(),
+                    execution_state: ExecutionState::restart_failed_no_retry(),
+                    ..Default::default()
```
[the remark this answers is not in the document]
Comment (inf17101) [reply agent/src/workload/workload_control_loop.rs:153]: I have seen this was fixed in previous commits.

File: server/src/ankaios_server.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (inf17101) [reply server/src/ankaios_server.rs]: Is fixed within previous commits.

File: agent/src/runtime_connectors/podman_kube/podman_kube_runtime.rs [hunk 16]
```diff
@@ -42,6 +42,16 @@ pub struct PodmanKubeWorkloadId {
     pub down_options: Vec<String>,
 }
 
+impl ToString for PodmanKubeWorkloadId {
+    fn to_string(&self) -> String {
+        if let Some(pods) = &self.pods {
+            sha256::digest(pods.join(""))
+        } else {
+            "no pods running".to_string()
```
[the remark this answers is not in the document]
Comment (krucod3) [reply agent/src/runtime_connectors/podman_kube/podman_kube_runtime.rs:50]: Done.

File: agent/src/runtime_connectors/podman/podman_runtime.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (krucod3) [reply agent/src/runtime_connectors/podman/podman_runtime.rs]: I also renamed the test now.

File: server/src/workload_state_db.rs (hunk 10 above)
Comment (krucod3) [reply server/src/workload_state_db.rs:23]: Done.

File: tests/resources/ankaios.resource (hunk 14 above)
Comment (krucod3) [reply tests/resources/ankaios.resource:184]: @lingnoi, maybe you can comment here. I think we can fix the newly introduced code, but would not do a major refactoring of the robot tests inside this PR.
If we have checks that could give a false PASS when used with an accepted argument, we should fix them to avoid always passing tests.

**Review by inf17101 — APPROVED** [review]
LGTM 👍

**Review by lingnoi — APPROVED** [review]
LGTM 👍
